contributors/design-proposals/volume-provisioning-configuration.md

```diff
@@ -0,0 +1,207 @@
+
+## Terminology
+
+* *admin* - user that can edit StorageClass objects.
+* *regular user* - user that can create PersistentVolumeClaims and cannot edit StorageClasses.
+
+## Use cases
+
+* Regular user wants get a PV of storage class "fast" in availability zone "east".
+* Admin wants to allow users to create PVs with storage class "fast" only in availability zones "east" and "north".
+* Regular user wants to discover what tunable properties for storage class "fast" are available and as result he can see that parameter "availability zone" with allowed values "east" and "north" is configurable.
+
+While availability zone is used as example in use cases, provisioners can choose any property such as encryption key, IOPS or replica count as tunable.
+
+## Design
+
+* Users can use `PersistentVolumeClaim.spec.provisioningOptions` with a pointer to `ConfigMap` in the same namespace. This `ConfigMap` contains options for the dynamic provisioning of a PV for this PVC.
+   ```go
+   type PersistentVolumeClaimSpec struct {
+      ...
+      // ProvisiongConfiguration contains a reference to ConfigMap with configuration of dynamic provisioning for this claim.
+      ProvisiongConfiguration *LocalObjectReference
+   }
+   ```
+   
+    * This way, PVCs are platform-independent and can be deployed elsewhere, as long as the user updates `ConfigMap(s)` with platform-dependent details of the provisioning.
+    * PV controller would ignore existing PVs and it would **always provision a new PV** for a PVC with non-empty `spec.provisioningOptions` - we do not have any way how to check if an existing PV conforms to `spec.provisioningOptions` so dynamic provisioning is the only way.
+    * All provisioners (both internal and external) must provision a PV that satisfies both `StorageClass.parameters` and `PVC.spec.provisioningOptions` or throw an error if such combination of options is not possible.
+    * When user does not specify any `provisioningOptions` or the referred `ConfigMap` does not specify all allowed parameters, the provisioner either uses some defaults or throws an error that some parameters in the ConfigMap are mandatory.
+
+* StorageClass will get a new field to tell regular users what are allowed options in `PVC.spec.provisioningOptions`:
+    ```go
+    type StorageClass struct {
+        // <snip>
+
+        // AllowedProvisioningOptions contains details about provisioning options that are allowed by this StorageClass.
+        AllowedProvisioningOption  []ProvisioningOptionSpec
+    }
+
+    // ProvisioningOption describes one option that can be used in PVC provisioningOptions
+    // when requesting a volume of a particular StorageClass.
+    type ProvisioningOptionSpec struct {
```

Comment:
this is a similar approach to how we handled Parameters in OpenShift templates, but we've found it generally ended up being insufficient / we kept needing to add fields to it to improve the experience.  In the open service broker API we decided instead to define parameters via JSON schema v4 since that will provide more flexibility and there are a plethora of JSON schema validators (and GUI form builders) out there.

Although from the perspective of the storage admin that is setting this up, it is more complex to define a JSON schema than a limited set of fields like is proposed here.